### PR TITLE
nanocoap: add coap_opt_put_uri_pathquery()

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

The URI is passed as a string to `suit_coap_get_blockwise()`.
It may contain query parameters (`/path/to/resource?foo=bar&baz=blub`), those will not be translated properly to CoAP options if `coap_opt_put_uri_path()` is used.

To fix this, add a function `coap_opt_put_uri_pathquery()` that handles both path and query parameters.

### Testing procedure

A test for the new function has been added to `tests/unittests/tests-nanocoap`.

### CODE REVIEW

**Comment by chrysn** [comment]
Given our CoAP messages are built for sequential option addition, I think that the better course here would be to parse the URI once, and then have the different functions to apply the components to a message.

For some applications it may be convenient to have a convenience function that sets all options, but that should be documented to be a convenience function equivalent to calling this particular list of functions, so that users who do need to add options inbetween can switch over easily.

**Comment by benpicco** [comment]
I agree, but I think for `suit_coap_get_blockwise()` this is not an option as it always expects a full URL.

File: sys/net/application_layer/nanocoap/nanocoap.c [hunk 1]
```diff
@@ -789,6 +787,24 @@ size_t coap_opt_put_string(uint8_t *buf, uint16_t lastonum, uint16_t optnum,
...
+    size_t bytes_out = coap_opt_put_string_with_len(buf, *lastonum,
+                                                    COAP_OPT_URI_PATH,
+                                                    uri, len, '/');
+    *lastonum = COAP_OPT_URI_PATH;
+
+    if (query) {
+        buf += bytes_out;
+        bytes_out += coap_opt_put_uri_query(buf, COAP_OPT_URI_PATH, query + 1);
+        *lastonum = COAP_OPT_URI_QUERY;
+    }
```
[the remark this answers is not in the document]
Comment (benpicco) [reply sys/net/application_layer/nanocoap/nanocoap.c]: Ok

**Comment by chrysn** [comment]
I've direct-pushed (hoped they'd show up as suggestions, but that seems to be yet another mechanism) a few changes on the documentation; mostly around terminology (the input is *not* a URI) and my original comment.

I'd prefer the function to be called `coap_opt_put_uri_pathquery` or similar -- the existing code is imprecise enough about what is a URI and what not, let's not add to that.

Personally I think that our string handling around URIs is a mess (who'd guess that `coap_option_put_string`'s precise handling of delimiters contains special provisions to make it just work for path components?), but this PR at least doesn't make it worse.

**Comment by benpicco** [comment]
I renamed the function, will change the commit message when squashing.

**Review by chrysn — APPROVED** [review]
Tested by running a SUIT update with the bogus `SUIT_COAP_BASEPATH='x?y'` for which the SUIT process then correctly tried to fetch a binary with Uri-Path x and Uri-Query y/...

ACK, please squash.

**Pull request merged**